### DISCUSSION TITLE
`dunai`: Re-structure API: `InternalCore` and `Util`

### DISCUSSION THREAD

**Discussion by ivanperez-keera** [ivanperez-keera/dunai#354] [Ideas]
An issue that has been bugging me for some time is that there are multiple modules at the same level in dunai with disparate functionality. currently, we publish:

```
        Data.MonadicStreamFunction.Async
        Data.MonadicStreamFunction.Core
        Instances
            Data.MonadicStreamFunction.Instances.ArrowChoice
            Data.MonadicStreamFunction.Instances.ArrowLoop
            Data.MonadicStreamFunction.Instances.ArrowPlus
            Data.MonadicStreamFunction.Instances.Num
            Data.MonadicStreamFunction.Instances.VectorSpace
        Data.MonadicStreamFunction.InternalCore
        Data.MonadicStreamFunction.Parallel
        Data.MonadicStreamFunction.ReactHandle
        Data.MonadicStreamFunction.Util
```

These modules can roughly be grouped as follows:

1) Core
```
        Data.MonadicStreamFunction.Core
        Data.MonadicStreamFunction.InternalCore
```

2) Instances of known classes

```
        Instances
            Data.MonadicStreamFunction.Instances.ArrowChoice
            Data.MonadicStreamFunction.Instances.ArrowLoop
            Data.MonadicStreamFunction.Instances.ArrowPlus
            Data.MonadicStreamFunction.Instances.Num
            Data.MonadicStreamFunction.Instances.VectorSpace
```

3) Extensions that do not need to import the internal core
```
        Data.MonadicStreamFunction.Util
```

4) Extensions that need to import the internal core
```
        Data.MonadicStreamFunction.Async
        Data.MonadicStreamFunction.Parallel
        Data.MonadicStreamFunction.ReactHandle
```

To me, the fact that they are all listed at the same level makes following all of this hard. I believe one has to keep these groups in their head and, although it's possible, it should be obvious to the user and the maintainer, to minimize the mental burden of using dunai.

In particular, the fact that `Core` and `InternalCore` are separate, and the fact that everything under number (4) is scattered around, just adds noise to the API.

The first issue has a solution: rename `Data.MonadicStreamFunction.InternalCore` to `Data.MonadicStreamFunction.Core.Internal`. That's also fairly common practice.

I can't think of a way to group everything under (4) that is not ugly. I'd welcome ideas in this regard. 

We do not want to put everything in `Util` directly. Importing the internal core should generally be avoided, and we are likely to get lazy and take shortcuts if we open that Pandora's box. It's useful to keep those separate (and, if we can, re-implement them without `InternalCore`), and it is telling us something about MSFs themselves.

A possibility would be to put everything under `Util`, and re-export it from there. To avoid circular imports (`Async` imports `Util`), we could just split `Util` into smaller modules.

There's also a chance that we need to split dunai into two or even three separate libraries (e.g., `dunai-core`, `dunai` proper, and `dunai-contrib`).

EDIT: Re-sort paragraphs.

**Comment by ivanperez-keera** [comment]
It could also make sense to have all those separate modules without `Util`. For example, `Data.MonadicStreamFunction.Fold`, `Data.MonadicStreamFunction.Delay`, and so on.

**Comment by ivanperez-keera** [comment]
What I like about this second way is that, instead of `Util` being a potpourri of everything we can do, if done well, the different modules would be driven by a top-level design view of what a streaming library should provide (based more on the underlying mathematical definition of an MSF, rather than utility).

**Comment by geroldmeisinger** [comment]
1.-3. makes sense to me

`Util` sounds like something I'm going to need all the time whereas `Async, Parallel and ReactHandle` sound like something advanced which I only need in special cases.

> Extensions that need to import the internal core

If they require `InternalCore`, to me it that would convey they are "complete" and there will be no more like these and everything else can be build based on `Core` and these "extras".

> I can't think of a way to group everything under (4) that is not ugly. I'd welcome ideas in this regard.

```
Data.MonadicStreamFunction.Async (concatS)
Data.MonadicStreamFunction.Parallel ((*|*), (&|&))
Data.MonadicStreamFunction.ReactHandle (ReactHandle, reactInit, react)
```

We are talking about 6 items in total split up into 4 groups. Is this hierarchy really necessary? Why not put them all in one package `Data.MonadicStreamFunction.CoreExtras`?

> There's also a chance that we need to split dunai into two or even three separate libraries

I don't think dunai is large or bloated enough (and probably won't be in the near future) to justify a separate library which would make everything scattered and hard to get an overview of.

**Comment by ivanperez-keera** [comment]
> We are talking about 6 items in total split up into 4 groups.

For now.

> Is this hierarchy really necessary? Why not put them all in one package Data.MonadicStreamFunction.CoreExtras?

That's the idea behind putting them under `Util` as separate modules and then re-export from `Util`.

From the point of view of the user, the distinction between `CoreExtras` and `Util` wouldn't be intuitive at all. The design needs to be logically organized and make sense to the user.

> I don't think dunai is large or bloated enough (and probably won't be in the near future) to justify a separate library which would make everything scattered and hard to get an overview of.

There are other criteria that are also important and that I'm trying to take into account here. One of them is to use libraries to delimit layers of abstraction. Having a good design (including a good separation into libraries) will decrease the maintenance burden and facilitate understanding dunai as a whole.
